### PR TITLE
WIP: Enhanced Categorical Accessor

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/dask/dask/issues/1573

I'm still working one this, but would appreciate some feedback
on the general approach. I'm touching some base clases here,
and don't want to break anything elsewhere.

### CODE REVIEW

File: dask/dataframe/tests/test_categorical.py [hunk 1]
```diff
@@ -78,3 +86,36 @@ def test_categories():
...
+        ('as_ordered', {}),
+        ('as_unordered', {}),
+        ('as_ordered', {}),
+        ('remove_categories', dict(removals=['a'])),
+        ('remove_unused_categories', {}),
+        ('rename_categories', dict(new_categories=['d', 'e', 'f'])),
+        ('reorder_categories', dict(new_categories=['a', 'b', 'c'])),
+        ('set_categories', dict(new_categories=['a', 'e', 'b'])),
+    ])
+    def test_callable(self, cat_series, method, kwargs):
```
Comment (TomAugspurger) [inline dask/dataframe/tests/test_categorical.py]: Thoughts on this kind of testing? It's pretty abstract, not easy to see exactly what's going on at first. I can certainly make these more concert if you like.

Comment (mrocklin) [inline dask/dataframe/tests/test_categorical.py]: I'm fine with it personally.  Covering the Pandas API comprehensively is hard.  Using parametrize seems like a good match.

**Comment by TomAugspurger** [comment]
I just pushed a commit with a few updates. Now the tests will also assert that the metadata (`._meta.cat.categories` and `._meta.cat.ordered`) are correct. This turned up a couple issues

1. My original implementation was broken :) I have to override `Accessor._function_map` to get the metadata to pass through correctly. Open to suggestions on how to do this better; there's a bit of code duplication.
2. Some operations can't reliably get the metadata correct.  For example `.remove_unused_categories` would have to do a full scan to know which categories are actually used; At the moment I've just raised an error on that. Does dask have a policy here? We can either (a) do a full scan, getting the metadata correct, (b) Allow the metadata to maybe be incorrect (c) raise an error. For my use case, being able to rely on the metadata is essential.

**Comment by mrocklin** [comment]
> Some operations can't reliably get the metadata correct. For example .remove_unused_categories would have to do a full scan to know which categories are actually used; At the moment I've just raised an error on that. Does dask have a policy here? We can either (a) do a full scan, getting the metadata correct, (b) Allow the metadata to maybe be incorrect (c) raise an error. For my use case, being able to rely on the metadata is essential.

That's tricky.  I don't think we have a strict policy here.  I also don't have any particular strong thoughts here.  We've managed to avoid this issue so far in most cases.  Perhaps we develop a convention of a keyword argument like `something='strict'`?

**Comment by TomAugspurger** [comment]
Thinking through this a bit more, it's not just the metadata that's a problem. There isn't really any way to do `remove_unused_categoricals` block-wise without reimplementing it in dask, since each block could have different present and missing values. That wouldn't be too difficult, but the workaround for the user isn't the worst

```python
In [2]: s = pd.Series(pd.Categorical(['a', 'b', 'b', 'a', 'b'], categories=['a', 'b', 'c']))

In [3]: a = dd.from_pandas(s, npartitions=2)

In [4]: present = a.unique().compute()

In [5]: a.cat.remove_categories(a.cat.categories.difference(present))
Out[5]: dd.Series<call-62..., npartitions=2, divisions=(0, 3, 4)>
```

**Comment by mrocklin** [comment]
If we didn't care about metadata, and if we don't care about low-memory operation (such as is true in the distributed case) then we would compute the categories of each partition, merge them somehow and then use that as an input to an embarrassingly parallel remove_categories call.  This is exactly what you've done above but encoded into the graph.  It has the unfortunate side effect of keeping all the data in memory afterwards if you're on a single machine.

We could also just call `.compute` within the Dask function.  We do this in a few other places.

**Comment by TomAugspurger** [comment]
I'm happy to implement whatever, but I don't have a good feeling for what's correct in this case :)

**Comment by mrocklin** [comment]
Correct behavior probably depends on choice of scheduler.   

I recommend that we optimize for the single-machine case (which is more conservative anyway) and trigger a compute call within `remove_unused_categoricals`.  This would involve doing pretty much exactly what you did above within the dask.dataframe function.

If we wanted to go all-out we could also add a `compute=True` keyword argument that would optionally do things lazily.

However I would be *very* happy to be overruled here if anyone has an opinion.  I don't use this functionality personally so I don't have much expertise on what's right here.

**Comment by TomAugspurger** [comment]
The last commit implements `remove_unused_categories`; the set of present categories is computed eagerly. I haven't had a chance to implement the `compute` keyword, and have a chance until later in the week probably.

**Comment by mrocklin** [comment]
This seems reasonable to me.  It looks like there are some merge conflicts.  I apologize for letting this linger.

**Comment by TomAugspurger** [comment]
No worries, I imagine Strata kept you busy :)

Just to make sure, you're OK with not implementing the `compute=True` keyword for avoiding the full scan on `remove_unused_categories`? I can always implement it later if it becomes a problem.

**Comment by mrocklin** [comment]
Yeah, this seems like a good step that doesn't stop us from making any good steps in the future.  Seems like a clear win to me.

**Comment by TomAugspurger** [comment]
Does dask prefer "Squash and Merge" or "Rebase and Merge"?

**Comment by jcrist** [comment]
Either or. I'd say use your best judgment has to whether the individual commits are worth keeping, or if squashing makes more sense. (Is there a difference for only a 1 commit PR?)

**Comment by mrocklin** [comment]
I've been squash-and-merging by default unless the commits are particularly well arranged.  In the cases where I've merged multiple commits I've tended to rebase them down cleanly ahead of time.  I usually only do this when a PR has multiple distinct features or when there are a couple of contributors and I want to preserve both of their activities.

**Pull request merged**